### PR TITLE
Hydra bridge v1 prototype

### PR DESCRIPTION
**Pull request by rimatik**
Here is all code regarding hydra bridge prototype

### CODE REVIEW

**Comment by AlexITC** [comment]
I'm leaving my comments for potential improvements to the app, I must say that the application is decent in many areas but I'm focusing only on the improvements, these will help to maintain the app when it grows.

General:
- There is no README, include the necessary details for running the apps, it would be ideal to get a section for running them in prod.
- There doesn't seem to be a module for sharing code between frontend/backend.
- We could create a `.nvmrc` file to specify the node version where the apps are tested.
- There isn't any single test.
- Missing linter.
- Missing code formatting rules.
- Missing CI integration.

Database:
- The postgres schema doesn't have consistent names, some use (example `Chain`/`chainId`/`created_at`/`Chain_pkey`).
- `SERIAL` is commonly discouraged, `UUID` usually works reasonably well.
- Avoid `TIMESTAMP`, use `TIMESTAMPTZ` instead.
- Mixed usage of `VARCHAR` and `TEXT`, `TEXT` should be good enough for most cases.
- It would be ideal to add comments about the purpose for non-obvious columns, for example `is_l1` likely means `is_layer1`, you can be specific in the names.

Backend:
- The service layer is too complex, it handles controller and database responsibilities, I'd suggest splitting this.
- Logs and error handling is very poor.
- While the route handlers have some input validations, they are usually quite basic.
- Mixes styles, when using `const`/`let`, you don't usually use `var`, but, there are parts of the app using `const`/`var`.
- It would be ideal to define the configuration in its own module, which would prevent from propagating the configuration loading to many places (e.g. `process.env.NODE_ENV || "dev";`)
- TypeScript semicolons aren't necessary in most cases
- There are huge functions which could easily get broken into smaller ones.
- REST APIs do not return useful error messages.

Frontend:
- Web app title needs to be updated from "React App"
- It is ideal to not surpress warnings, like `// eslint-disable-line react-hooks/exhaustive-deps`, these warnings are there for a good reason.
- `useHome` is a surprisingly big hook, I'd suggest to avoid this pattern, it also returns quite a lot of items.
- It would be ideal to define the configuration in its own module, which would prevent from propagating the configuration loading to many places (e.g. `process.env.NODE_ENV || "dev";`)
- `Home` is a huge component which can be broken down into many components, it is hard to follow it.
- Try to avoid using `any`, there are many places where it seems unnecessary, for example, `useHome` has this piece `const [error, setError] = useState<any>(undefined);`
- The application feels really slow
- When you load the app, the `Connect Wallet` button has a gray background, which makes me think it isn't enabled but it is..
- When you load the app, `Connect Wallet` button must be clicked, it would be ideal to make this requirement clear, I'm not sure if it makes sense to leave the calculator enablen when the wallet isn't connected.
- `Available routes: ` is displayed when there is no route, I'd choose a different message.
- The progress indicators are pretty useful, it would be ideal to display an estimator on how long the operation should take.

Smart contracts:
- Overall, I'd have expected many tests because this is a critical piece but there is none.

Thanks!

File: backend/api/services/commonService.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+          chain: {
+            select: {
+              id: true,
+              chainId: true,
+            },
+          },
+        },
+      });
+      const tokens: TokenResponseDto[] = [];
+      for (let i = 0; i < chainTokens.length; i++) {
```
Comment (iBlackShadow) [inline backend/api/services/commonService.ts:57]: this could be a map instead of manually building the new array

File: backend/api/services/commonService.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            decimals: true,
+            symbol: true,
+          },
+        },
+      },
+    });
+
+    if (chains) {
+      const chainsResponse: ChainResponseDto[] = [];
+      for (let i = 0; i < chains.length; i++) {
```
Comment (iBlackShadow) [inline backend/api/services/commonService.ts:101]: this could be a map instead of manually building the new array

File: backend/app.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  ignoreRoute: function (req, res) { return false; }
+}));
+
+
+app.use("/api", routes);
+
+app.use("/swagger", swaggerUI.serve, swaggerUI.setup(swaggerDoc));
+
+app.listen(PORT, () => {
+  console.log("\x1b[34m", `Host: ${URL}:${PORT}/api/v${VERSION}`);
```
Comment (iBlackShadow) [inline backend/app.ts:37]: i believe the printed host is wrong, it should be `${URL}:${PORT}/api` and the VERSION env variable is not really being used

File: backend/hardhat/contracts/HydraBridge.sol [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "./interfaces/polygon/IRootChainManager.sol";
+import "./interfaces/hop/IL1_Bridge.sol";
+
+contract HydraBridge is Ownable, Pausable {
```
Comment (iBlackShadow) [inline backend/hardhat/contracts/HydraBridge.sol:12]: i dont think Pausable is being really used at the moment, only the pause and unpause methods check if the contract is paused or not

File: backend/api/services/bridgeService.ts [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    );
+
+    return {
+      data: sendToL2HopData,
+      to: ETH_CONTRACT,
+      from: dto.recipient,
+    };
+  }
+
+  return {
```
Comment (iBlackShadow) [inline backend/api/services/bridgeService.ts:344]: no support for ETH on hop?, i see that the smart contract has a `sendEthToL2Hop` method

File: frontend/src/modules/Home/Home.tsx [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                isEth={isEth}
+                isAmountSet={!!amountIn}
+                isAbleToMove={isAbleToMove}
+                onWalletConnect={onConnectWallet}
+                onWalletApprove={onApproveWallet}
+                onMoveAssets={handleMoveAssets}
+              />
+            </Container>
+          </TransferWrapper>
+          {isAbleToMove && !!amountIn && amountIn > 0 && isConnected && (
```
Comment (iBlackShadow) [inline frontend/src/modules/Home/Home.tsx:323]: i think it would be nice for the users to see the available routes before calling approve and paying a fee

File: frontend/src/common/components/BridgeRoutes/BridgeRoutes.tsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+            return (
+              <Route
+                key={route.id}
+                isSelected={isSelected}
+                onClick={() => onClick(route.id)}
+              >
+                <RouteContent>
+                  <FeeContainer>
+                    {"Fee: $" +
```
Comment (iBlackShadow) [inline frontend/src/common/components/BridgeRoutes/BridgeRoutes.tsx:205]: it would be helpful to also see the amount of ETH paid as fee

File: frontend/src/common/components/BridgeRoutes/BridgeRoutes.tsx [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+        </SpinnerContainer>
+      ) : (
+        <Container>
+          {bridgeRoutes.map((route: RouteCalculatedDto) => {
+            const assetIconName =
+              route.bridgeRoute.fromAsset.symbol.toLocaleLowerCase() as IconKeys;
+            const isSelected = selectedRouteId === route.id;
+
+            const formatAmountIn =
+              route.bridgeRoute.amountIn !== ""
```
Comment (iBlackShadow) [inline frontend/src/common/components/BridgeRoutes/BridgeRoutes.tsx:185]: this could be a function to avoid duplication the same code to format the amountOut
